Add documentation for process.env, process.nextTick, and diagnostics_channel (#9571)
cloudflare/cloudflare-docs: content/workers/runtime-apis/nodejs/process.md

---
pcx_content_type: configuration
title: process
---

# process

{{<render file="_nodejs-compat-howto.md">}}

> The `process` module in Node.js provides a number of useful APIs related to the current process.

Reviewer: ```suggestion
The [`process`](https://nodejs.org/dist/latest-v19.x/docs/api/process.html) module in Node.js provides a number of useful APIs related to the current process.
```

> Within the Workers environment, most of these APIs are not relevant or meaningful, but some are

Reviewer: ```suggestion
Within a serverless environment like Workers, most of these APIs are not relevant or meaningful, but some are
```

useful for cross-runtime compatibility. Within workers the following APIs are available:

```js
import {
  env,
  nextTick,
} from 'node:process';

env['FOO'] = 'bar';
console.log(env['FOO']); // Prints: bar

nextTick(() => {
  console.log('next tick');
});
```

> In the Node.js implementation of `process.env`, the `env` object is a copy of the environment

Author: ```suggestion
## `process.env`

In the Node.js implementation of `process.env`, the `env` object is a copy of the environment
```

variables at the time the process was started. In the Workers implementation, there is no
process level environment, so `env` is an empty object. You can still set and get values from
`env`, and those will be globally persistent for all workers running in the same Isolate and
context (e.g. the same script).

> Workers do have a concept of [environment variables](/workers/platform/environment-variables/) that are applied on a per-worker/per-request basis. These are not accessible via the `process.env` API.

Reviewer: ```suggestion
Workers do have a concept of [environment variables](/workers/platform/environment-variables/) that are applied on a per-worker/per-request basis. These are not accessible via the `process.env` API, but if necessary you can assign environment variables to `process.env` as shown below:

```typescript
export default { 
  fetch(request, env) {
    process.env = env;
    // ...
  }
}
```

```

Maybe help people connect dots on how to do this? (even if it sounds obvious)

Author: Assigning the request level `env` to `process.env` is a bit dangerous given that (a) `process.env` is global and used for all workers using the same script and (b) the request level `env` contains more than just strings and is generally specific to that request's `IoContext`. Which is a long way of saying I *don't* think we want to encourage folks to do this kind of assignment.

Also, because of the way ESM imports work, assigning `process.env` to a new value like this would NOT change the value of `env` as a named export... that is,

```
import * as process from 'node:process';
import { env } from 'node:process';

console.log(env === process.env); // true! these are the same object

export default {
  fetch(request, env) {
    process.env = env;
    console.log(env === process.env);  // false!
    process.env['FOO'] = 'abc';
    console.log(env['FOO'] === process.env['FOO']);  // false! these are different objects now.
  }
};

```

Reviewer: Then maybe an example that shows assigning a specific key on `env` to `process.env`

```javascript
import * as process from 'node:process';

export default {
  fetch(request, env) {
    process.env.FOO = env.FOO;
    // ...
  }
}
```

Just feels like we'd want to show people _some_ way to set this

Author: Added

> The Workers implementation of `process.nextTick()` is a wrapper for the standard Web Platform API

Reviewer: Add heading for this section

Author: ```suggestion
## `process.nextTick()`

The Workers implementation of `process.nextTick()` is a wrapper for the standard Web Platform API
```

---
pcx_content_type: configuration
title: process
---

# process

{{<render file="_nodejs-compat-howto.md">}}

> The [`process`](https://nodejs.org/dist/latest-v19.x/docs/api/process.html) module in Node.js provides a number of useful APIs related to the current process.
> Within a serverless environment like Workers, most of these APIs are not relevant or meaningful, but some are
> useful for cross-runtime compatibility. Within workers the following APIs are available:

Reviewer 2: ```suggestion
The [`process`](https://nodejs.org/dist/latest-v19.x/docs/api/process.html) module in Node.js provides a number of useful APIs related to the current process. Within a serverless environment like Workers, most of these APIs are not relevant or meaningful, but some are useful for cross-runtime compatibility. Within Workers, the following APIs are available:
```

```js
import {
  env,
  nextTick,
} from 'node:process';

env['FOO'] = 'bar';
console.log(env['FOO']); // Prints: bar

nextTick(() => {
  console.log('next tick');
});
```

## `process.env`

> In the Node.js implementation of `process.env`, the `env` object is a copy of the environment
> variables at the time the process was started. In the Workers implementation, there is no
> process level environment, so `env` is an empty object. You can still set and get values from
> `env`, and those will be globally persistent for all workers running in the same Isolate and
> context (e.g. the same script).

Reviewer 2: ```suggestion
In the Node.js implementation of `process.env`, the `env` object is a copy of the environment variables at the time the process was started. In the Workers implementation, there is no process-level environment, so `env` is an empty object. You can still set and get values from `env`, and those will be globally persistent for all Workers running in the same isolate and context (for example, the same Workers entry point).
```

### Relationship to per-request `env` argument in `fetch()` handlers

> Workers do have a concept of [environment variables](/workers/platform/environment-variables/) that
> are applied on a per-worker/per-request basis. These are not accessible automatically via the
> `process.env` API. It is possible to manually copy these values into `process.env` if you need
> to. Be aware, however, that setting any value on `process.env` will coerce that value into a
> string.

Reviewer 2: ```suggestion
Workers do have a concept of [environment variables](/workers/platform/environment-variables/) that are applied on a per-Worker and per-request basis. These are not accessible automatically via the `process.env` API. It is possible to manually copy these values into `process.env` if you need to. Be aware, however, that setting any value on `process.env` will coerce that value into a string.
```

```js
import * as process from 'node:process';

export default {
  fetch(req, env) {
    // Set process.env.FOO to the value of env.FOO if process.env.FOO is not already set
    // and env.FOO is a string.
    process.env.FOO ??= (() => {
      if (typeof env.FOO === 'string') {
        return env.FOO;
      }
    })();
  }
};
```

> It is strongly recommended that you *do not* replace the entire `process.env` object with the
> request `env` object. Doing so will cause you to lose any environment variables that were set
> previously and will cause unexpected behavior for other workers running in the same Isolate.
> Specifically, it would cause inconsistency with the `process.env` object when accessed via
> named imports.

Reviewer 2: ```suggestion
It is strongly recommended that you *do not* replace the entire `process.env` object with the request `env` object. Doing so will cause you to lose any environment variables that were set previously and will cause unexpected behavior for other Workers running in the same isolate. Specifically, it would cause inconsistency with the `process.env` object when accessed via named imports.
```

```js
const * as process from 'node:process';
const { env } from 'node:process';

process.env === env; // true! they are the same object
process.env = {}; // replace the object! Do not do this!
process.env === env; // false! they are no longer the same object

// From this point forward, any changes to process.env will not be reflected in env,
// and vice versa!
```

## `process.nextTick()`

> The Workers implementation of `process.nextTick()` is a wrapper for the standard Web Platform API
> [`queueMicrotask()`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/queueMicrotask).

Reviewer 2: ```suggestion
The Workers implementation of `process.nextTick()` is a wrapper for the standard Web Platform API [`queueMicrotask()`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/queueMicrotask).
```
